xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -0,0 +1,725 @@
+/*
+ * Copyright 2019 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Stopwatch;
+import com.google.common.collect.ImmutableList;
+import com.google.protobuf.InvalidProtocolBufferException;
+import io.envoyproxy.envoy.api.v2.DiscoveryRequest;
+import io.envoyproxy.envoy.api.v2.DiscoveryResponse;
+import io.envoyproxy.envoy.api.v2.Listener;
+import io.envoyproxy.envoy.api.v2.RouteConfiguration;
+import io.envoyproxy.envoy.api.v2.core.Node;
+import io.envoyproxy.envoy.api.v2.route.Route;
+import io.envoyproxy.envoy.api.v2.route.VirtualHost;
+import io.envoyproxy.envoy.config.filter.network.http_connection_manager.v2.HttpConnectionManager;
+import io.envoyproxy.envoy.config.filter.network.http_connection_manager.v2.Rds;
+import io.envoyproxy.envoy.service.discovery.v2.AggregatedDiscoveryServiceGrpc;
+import io.grpc.ManagedChannel;
+import io.grpc.ManagedChannelBuilder;
+import io.grpc.Status;
+import io.grpc.SynchronizationContext;
+import io.grpc.SynchronizationContext.ScheduledHandle;
+import io.grpc.alts.GoogleDefaultChannelBuilder;
+import io.grpc.internal.BackoffPolicy;
+import io.grpc.stub.StreamObserver;
+import io.grpc.xds.Bootstrapper.ChannelCreds;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+
+final class XdsClientImpl extends XdsClient {
```

Comment:
Can we split this class into multiple smaller classes for ease of maintenance and review? For example, `RouteConfigurationProcessor` for all `RouteConfiguration` processing logic and so on.

Reply:
That only delegates the complexity to `RouteConfigurationProcessor` and you call a complex method. No benefits for doing that.

Reply:
Actually that's exactly the benefit - delegating to another class some of the complexity so that this class is easier to review/maintain etc.

Reply:
Somewhat unrelated comment: this class could also be used from outside this package (from `sds`) so we should look into making this public but mark `@Internal` . This also

Reply:
> Actually that's exactly the benefit - delegating to another class some of the complexity so that this class is easier to review/maintain etc.

We've had a long conversation with @zhangkun83 about having a monolithic class vs. multiple subcomponents twisting together, and this is the approach we decided to proceed. This class is big, but not that complex.

Reply:
> Somewhat unrelated comment: this class could also be used from outside this package (from `sds`) so we should look into making this public but mark `@Internal` . Also an argument for simplifying this class...

Do you need a reference to it from another package now? If not, we would be focusing on gRPC's load balancing usage for now and do this later.

Reply:
This is not urgent - just an FYI